src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -0,0 +1,3410 @@
+ * different replicas (e.g. of a reference table), which could lead the
+ * replicas to diverge.
+ *
+ * The second case prevents deadlocks due to out-of-order execution.
+ *
+ * We do not take executor shard locks for utility commands such as
+ * TRUNCATE because the table locks already prevent concurrent access.
+ */
+static void
+AcquireExecutorShardLocks(DistributedExecution *execution)
+{
+	CmdType operation = execution->operation;
+	List *taskList = execution->tasksToExecute;
+
+	if (!(operation == CMD_INSERT || operation == CMD_UPDATE || operation == CMD_DELETE ||
+		  SelectForUpdateOnReferenceTable(operation, taskList)))
+	{
+		/*
+		 * Executor locks only apply to DML commands and SELECT FOR UPDATE queries
+		 * touching reference tables.
+		 */
+		return;
+	}
+
+	/*
+	 * When executing in sequential mode or only executing a single task, we
+	 * do not need multi-shard locks.
+	 */
+	if (list_length(taskList) == 1 || ShouldRunTasksSequentially(taskList))
+	{
+		ListCell *taskCell = NULL;
+
+		foreach(taskCell, taskList)
+		{
+			Task *task = (Task *) lfirst(taskCell);
+
+			AcquireExecutorShardLock(task, operation);
+		}
+	}
+	else if (list_length(taskList) > 1)
+	{
+		AcquireExecutorMultiShardLocks(taskList);
+	}
+}
+
+
+/*
+ * FinishDistributedExecution cleans up resources associated with a
+ * distributed execution. In particular, it releases connections and
+ * clears their state.
+ */
+static void
+FinishDistributedExecution(DistributedExecution *execution)
+{
+	UnsetCitusNoticeLevel();
+
+	if (DistributedExecutionModifiesDatabase(execution))
+	{
+		/* prevent copying shards in same transaction */
+		XactModificationLevel = XACT_MODIFICATION_DATA;
+	}
+}
+
+
+/*
+ * CleanUpSessions does any clean-up necessary for the session
+ * used during the execution.
+ */
+static void
+CleanUpSessions(DistributedExecution *execution)
+{
+	List *sessionList = execution->sessionList;
+	ListCell *sessionCell = NULL;
+
+	/* always trigger wait event set in the first round */
+	foreach(sessionCell, sessionList)
+	{
+		WorkerSession *session = lfirst(sessionCell);
+
+		MultiConnection *connection = session->connection;
+		RemoteTransaction *transaction = &(connection->remoteTransaction);
+		RemoteTransactionState transactionState = transaction->transactionState;
+
+		elog(DEBUG4, "Total number of commands sent over the session %ld: %ld",
+			 session->sessionId, session->commandsSent);
+
+		UnclaimConnection(connection);
+
+		if (connection->connectionState == MULTI_CONNECTION_CONNECTING ||
+			connection->connectionState == MULTI_CONNECTION_FAILED)
+		{
+			/*
+			 * We want the MultiConnection go away and not used in
+			 * the subsequent executions.
+			 */
+			CloseConnection(connection);
+		}
+		else if (transactionState == REMOTE_TRANS_CLEARING_RESULTS)
+		{
+			/* TODO: should we keep this in the state machine? or cancel? */
+			ClearResults(connection, false);
+		}
+
+		/* TODO: can we be in SENT_COMMAND? */
```

Comment:
Do you think we should do anything about the TODO's here?

Reply:
I don't think so, I think I'll push a commit to remove those comments unless you have further  comments.

The reason the execution could only finish:
    - Connection State of `MULTI_CONNECTION_CONNECTING` or `MULTI_CONNECTION_FAILED` or `MULTI_CONNECTION_CONNECTED`.
    - In the first `if` statement we handle `MULTI_CONNECTION_CONNECTING` and `MULTI_CONNECTION_FAILED` 
    - If the state is  `MULTI_CONNECTION_CONNECTED` the `TransactionStateMachine` ensures that we're on `REMOTE_TRANS_STARTED` or `REMOTE_TRANS_INVALID` in the end of the state transitions (note that on any failures/unexpected things we'd be back on `MULTI_CONNECTION_FAILED` state)

Reply:
(also, check-full doesn't hit anything other lines/states that I explained above.)

Reply:
Hmm, wait, the transaction might be in `REMOTE_TRANS_CLEARING_RESULTS` because if the session has sent `BEGIN` but there is not enough tasks to assign to that, and so never actually got to the point where it is fully cleared. So, we need to keep the  `REMOTE_TRANS_CLEARING_RESULTS`. 

Do you see any other such edge cases?